### PR TITLE
[tapocontrol] now supports energy monitoring of P110 device

### PR DESCRIPTION
**Pull request by wildcs**
Added "EnergyMonitoring" function for P110 SmartPlugs

Device will report:
* actual PowerUsage (Watt)
* today EnergyUsage (Wh)
* today Runtime (Time in seconds device was on today)

### CODE REVIEW

File: bundles/org.openhab.binding.tapocontrol/README.md [hunk 1]
```diff
@@ -87,6 +96,9 @@ All devices support some of the following channels:
 |           | color            | Color                  | Color                        | L530, L900                                  |
 | device    | wifiSignal       | system.signal-strength | WiFi-quality-level           | P100, P105, P110, L510, L530, L900, L920    |
 |           | onTime           | Number:Time            | seconds output is on         | P100, P105, P110, L510, L530, L900, L920    |
+| energy    | actualPower      | Number:Power           | actual Power (Watt)          | P110                                        |
+|           | todayEnergyUsage | Number:Energy          | used energy today (Wh)       | P110                                        |
+|           | todyRuntime      | Number:Time            | seconds output was on today  | P110                                        |
```
Comment (lolodomo) [inline bundles/org.openhab.binding.tapocontrol/README.md:101]: Typo: tod**a**yRuntime

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/constants/TapoThingConstants.java [hunk 2]
```diff
@@ -143,6 +155,11 @@
     public static final String CHANNEL_ONTIME = "onTime";
     public static final String CHANNEL_OVERHEAT = "overheated";
     public static final String CHANNEL_WIFI_STRENGTH = "wifiSignal";
+    // channel group energy monitor
+    public static final String CHANNEL_GROUP_ENERGY = "energy";
+    public static final String CHANNEL_NRG_POWER = "actualPower";
+    public static final String CHANNEL_NRG_USAGE_TODAY = "todayEnergyUsage";
+    public static final String CHANNEL_NRG_RUNTIME_TODAY = "todyRuntime";
```
Comment (lolodomo) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/constants/TapoThingConstants.java:162]: Typo: tod**a**yRuntime

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/helpers/TapoUtils.java [hunk 3]
```diff
@@ -342,7 +344,29 @@ public static HSBType getHSBType(Integer hue, Integer saturation, Integer bright
      * @param unit TimeUnit (Unit<Time>)
      * @return QuantityTime<Time>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/helpers/TapoUtils.java:345]: QuantityType<Time>

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/helpers/TapoUtils.java [hunk 4]
```diff
@@ -342,7 +344,29 @@ public static HSBType getHSBType(Integer hue, Integer saturation, Integer bright
...
+    public static QuantityType<Time> getTimeType(@Nullable Number numVal, Unit<Time> unit) {
+        return new QuantityType<>((numVal != null ? numVal : 0), unit);
+    }
+
+    /**
+     * Return QuantityType with Power
+     * 
+     * @param numVal Number with value
+     * @param unit PowerUnit (Unit<Power>)
+     * @return QuantityTime<Power>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/helpers/TapoUtils.java:356]: QuantityType<Power>

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/helpers/TapoUtils.java [hunk 5]
```diff
@@ -342,7 +344,29 @@ public static HSBType getHSBType(Integer hue, Integer saturation, Integer bright
...
+    public static QuantityType<Power> getPowerType(@Nullable Number numVal, Unit<Power> unit) {
+        return new QuantityType<>((numVal != null ? numVal : 0), unit);
+    }
+
+    /**
+     * Return QuantityType with Energy
+     * 
+     * @param numVal Number with value
+     * @param unit PowerUnit (Unit<Power>)
+     * @return QuantityTime<Power>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/helpers/TapoUtils.java:367]: QuantityType<Energy>

File: bundles/org.openhab.binding.tapocontrol/src/main/resources/OH-INF/thing/channels.xml [hunk 6]
```diff
@@ -148,8 +159,36 @@
...
+		<category>Energy</category>
+		<state readOnly="true" pattern="%.2f %unit%"></state>
+	</channel-type>
+
+	<!-- today runtime -->
+	<channel-type id="todayRuntimeChannel">
+		<item-type>Number:Time</item-type>
+		<label>Today Runtime</label>
+		<description>Today runtime (On-Time)</description>
+		<category>Energy</category>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.tapocontrol/src/main/resources/OH-INF/thing/channels.xml:186]: Category looks inappropriate

File: bundles/org.openhab.binding.tapocontrol/src/main/resources/OH-INF/thing/channels.xml [hunk 7]
```diff
@@ -62,6 +62,17 @@
...
 	</channel-group-type>
 
+	<!-- Energy Monitor -->
+	<channel-group-type id="energyMonitor">
+		<label>Energy Usage</label>
+		<description>Energy and Power usage</description>
+		<channels>
+			<channel id="actualPower" typeId="actualPowerChannel"></channel>
+			<channel id="todayEnergyUsage" typeId="todayEnergyUsageChannel"></channel>
+			<channel id="todyRuntime" typeId="todayRuntimeChannel"></channel>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.tapocontrol/src/main/resources/OH-INF/thing/channels.xml:72]: Typo: tod**a**yRuntime

File: bundles/org.openhab.binding.tapocontrol/src/main/resources/OH-INF/thing/channels.xml [hunk 8]
```diff
@@ -148,8 +159,36 @@
...
+		<state readOnly="true" pattern="%.2f %unit%"></state>
+	</channel-type>
+
+	<!-- today runtime -->
+	<channel-type id="todayRuntimeChannel">
+		<item-type>Number:Time</item-type>
+		<label>Today Runtime</label>
+		<description>Today runtime (On-Time)</description>
+		<category>Energy</category>
+		<state readOnly="true" pattern="%s %unit%"></state>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.tapocontrol/src/main/resources/OH-INF/thing/channels.xml:187]: %s looks strange to me ! You have a number, it should be either %d or %f

File: bundles/org.openhab.binding.tapocontrol/src/main/resources/OH-INF/i18n/tapocontrol.properties [hunk 9]
```diff
@@ -77,3 +81,24 @@ channel-type.tapocontrol.outputChannel.label = Output Switch
...
+
+channel-group-type.tapocontrol.commandDebug.label = Device Communication Debug
+channel-group-type.tapocontrol.commandDebug.description = Device resoponses and command debugging
+
+# channel types
+
+channel-type.tapocontrol.deviceCommand.label = Device Command
+channel-type.tapocontrol.deviceCommand.description = command send to device. use: 'command':'value'
+channel-type.tapocontrol.deviceResponse.label = Device Response
+channel-type.tapocontrol.deviceResponse.description = DeviceResponse
```
Comment (lolodomo) [inline bundles/org.openhab.binding.tapocontrol/src/main/resources/OH-INF/i18n/tapocontrol.properties:104]: Why all these entries are now added while I do not see corresponding stuff added in your changes ?
Was it previously missing ?

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.tapocontrol/src/main/resources/OH-INF/i18n/tapocontrol.properties (hunk 9 above)
Comment (wildcs) [reply bundles/org.openhab.binding.tapocontrol/src/main/resources/OH-INF/i18n/tapocontrol.properties]: This were generated from testdevice in \src\test folder. I removed it in .properties now

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/helpers/TapoUtils.java [hunk 10]
```diff
@@ -340,9 +342,31 @@ public static HSBType getHSBType(Integer hue, Integer saturation, Integer bright
...
+    public static QuantityType<Power> getPowerType(@Nullable Number numVal, Unit<Power> unit) {
+        return new QuantityType<>((numVal != null ? numVal : 0), unit);
+    }
+
+    /**
+     * Return QuantityType with Energy
+     * 
+     * @param numVal Number with value
+     * @param unit PowerUnit (Unit<Power>)
+     * @return QuantityType<Power>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/helpers/TapoUtils.java:367]: QuantityType<Energy>

**Review by lolodomo — CHANGES_REQUESTED** [review]

**Review by lolodomo — APPROVED** [review]
LGTM

**Pull request merged**